File: worker-addon/ryanair.js
const init_iframe = () => {
    const iframe = document.createElement('iframe');
    iframe.style.display = 'none';
    document.body.appendChild(iframe);
    return iframe;
}

const execute_in_iframe = (iframe, func) => {
    const iframe_window = iframe.contentWindow;
    const iframe_document = iframe_window.document;
    const iframe_div = iframe_document.createElement('div');
    iframe_div.addEventListener('click', func);
    iframe_div.click();
};

const common_headers = {
    'Client': 'desktop',
    'Client-Version': '3.153.0',
};

const list_all_airports = async () => {
    const url = 'https://www.ryanair.com/api/views/locate/5/airports/en/active';

    const response = await fetch(url, {
        method: 'GET',
        headers: common_headers,
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText}}`);
    }
    return {
        airports: await response.json(),
        fetch_date: new Date().toISOString(),
    };
};

const find_dest_airpoits = async (src_code) => {
    const url = `https://www.ryanair.com/api/views/locate/searchWidget/routes/en/airport/${src_code}`;

    const response = await fetch(url, {
        method: 'GET',
        headers: common_headers,
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText}}`);
    }
    return {
        airports: await response.json(),
        src_code: src_code,
        fetch_date: new Date().toISOString(),
    };
};

const query_available_dates = async (src_code, dst_code) => {
    const url = `https://www.ryanair.com/api/farfnd/3/oneWayFares/${src_code}/${dst_code}/availabilities`

    const response = await fetch(url, {
        method: 'GET',
        headers: common_headers,
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText}}`);
    }
    return {
        dates: await response.json(),
        src_code: src_code,
        dst_code: dst_code,
        fetch_date: new Date().toISOString(),
    };
};

const query_flights_details = async (
    src_code, dst_code, date, opts
) => {
    const url = `https://www.ryanair.com/api/booking/v4/en-gb/availability`
        + `?ADT=${opts?.adult ?? 1}&TEEN=${opts?.teen ?? 0}&CHD=${opts?.child ?? 0}&INF=${opts?.infant ?? 0}`
        + `&Origin=${src_code}&Destination=${dst_code}&promoCode=&IncludeConnectingFlights=false`
        + `&DateOut=${date}&DateIn=`
        + `&FlexDaysBeforeOut=${opts?.days_before ?? 2}&FlexDaysOut=${opts?.days_after ?? 2}`
        + `&FlexDaysBeforeIn=${opts?.days_before ?? 2}&FlexDaysIn=${opts?.days_after ?? 2}`
        + `&RoundTrip=false&IncludePrimeFares=false&ToUs=AGREED`;

    const response = await fetch(url, {
        method: 'GET',
        headers: common_headers
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText}}`);
    }
    return {
        booking: await response.json(),
        src_code: src_code,
        dst_code: dst_code,
        date: date,
        fetch_date: new Date().toISOString(),
    };
};

const manager_url = 'http://localhost:8090'
const storage_url = `${manager_url}/storage`
const scheduler_url = `${manager_url}/scheduler`

const save_data = async (data) => {
    const url = `${storage_url}/save_result`;
    const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify({
            dataset_name: 'ryanair',
            result: data,
        }),
        mode: "cors",
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText} for ${url}`);
    }
};

const fetch_job = async () => {
    const url = `${scheduler_url}/fetch_job`;
    const response = await fetch(url, {
        method: 'POST',
        body: '',
        mode: "cors",
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText} for ${url}`);
    }
    return await response.json();
};

const complete_job = async (job_id) => {
    const url = `${scheduler_url}/complete_job`;
    const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify({
            job_id: job_id,
        }),
        mode: "cors",
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText} for ${url}`);
    }
};

const save_flight_dates = async (src_code, dst_code, dates) => {
    const url = `${scheduler_url}/save_flight_dates`;
    const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify({
            src_code: src_code,
            dst_code: dst_code,
            dates: dates,
        }),
        mode: "cors",
    });
    if (response.status !== 200) {
        throw new Error(`Server responded with status ${response.status} ${response.statusText} for ${url}`);
    }
};

const sleep = (base_s=2000, rand_s=5000) => new Promise((resolve) => setTimeout(resolve, base_s + Math.random() * rand_s))

const process_fetch_dates = async (src_code, dst_code) => {
    const dates = await query_available_dates(src_code, dst_code);
    await save_flight_dates(src_code, dst_code, dates.dates);
};

const process_fetch_details = async (src_code, dst_code, date) => {
    const details = await query_flights_details(src_code, dst_code, date);
    await save_data(details);
};

const process_job = async () => {
    const job_data = await fetch_job();
    console.log(job_data);
    if (!job_data) {
        console.log('No jobs available');
        return;
    }
    console.log('Fetched job', job_data);

    switch(job_data.type_) {
        case 'QueryDatesJob':
            await process_fetch_dates(job_data.src_code, job_data.dst_code);
            break;
        case 'QueryFlightsJob':
            await process_fetch_details(job_data.src_code, job_data.dst_code, job_data.date);
            break;
        default:
            throw new Error(`Unknown job type ${job_data.type}`);
    }

    await complete_job(job_data.id);
};

const worker_process = async () => {
    let allow_errors = 8;
    while (allow_errors > 0) {
        try {
            await process_job();
        } catch (e) {
            console.error('Job execution error', e);
            allow_errors--;
        }
        await sleep();
        await sleep();
    }
    window.close();
};

console.log('YEEEEAH!!!!! Worker started xD');
setTimeout(async () => {
    await worker_process();
    // console.log(await query_available_dates('WAW', 'MAN'));
    // console.log(await query_flights_details('WAW', 'MAN', '2025-07-10', { days_before: 2, days_after: 2 }));
}, 5000);

// TODO: simulate user behaviour

File: webdrv/flights_scrape_webdrv/ryanair.py
import logging
from functools import partial
from datetime import datetime, timedelta
from typing import Tuple, Optional
import websockets

from .utils import new_id, sleep_rand
from .webdrv import WSTab
from .store import save_result
from .job import Job, load_jobs, run_jobs


RYANAIR_API_URL = 'https://www.ryanair.com/api'


class QueryDatesJob(Job):
    src_code: str
    dst_code: str
    start_date: str
    end_date: str


class QueryFlightsJob(Job):
    src_code: str
    dst_code: str
    date: str
    days_before: int = 2
    days_after: int = 2


def make_dates_jobs(
    src_code: str, dst_code: str, dates: Tuple[str, ...],
    days_before: int = 2, days_after: int = 2
):
    dates = sorted(dates)
    while dates:
        date = dates.pop()
        date_parsed = datetime.strptime(date, '%Y-%m-%d')
        date_query = (date_parsed - timedelta(days=days_before)).strftime('%Y-%m-%d')
        date_start = (date_parsed - timedelta(days=days_before+days_after+1)).strftime('%Y-%m-%d')
        while dates and dates[-1] >= date_start:
            dates.pop()

        yield QueryFlightsJob(
            id=new_id(),
            src_code=src_code,
            dst_code=dst_code,
            date=date_query,
            days_before=days_before,
            days_after=days_after,
        )


def init_jobs(airports, start_date, end_date):
    for a1 in airports:
        for a2 in airports:
            if a1 != a2:
                yield QueryDatesJob(
                    id=new_id(),
                    src_code=a1,
                    dst_code=a2,
                    start_date=start_date,
                    end_date=end_date,
                )


RYANAIR_HEADERS = {
    'Client': 'desktop',
    'Client-Version': '3.153.0',
}


async def make_ryanair_request(tab: WSTab, url: str):
    print(f'Ryanair: {url}')
    msg_id = await tab.run_code('''(async (url, common_headers) => {
        const response = await fetch(url, {
            method: 'GET',
            headers: common_headers,
        });
        if (response.status !== 200) {
            throw new Error(`Server responded with status ${response.status}`);
        }
        return {
            result: await response.json(),
            fetch_date: new Date().toISOString(),
        };
    })'''
    f'''('{url}', {RYANAIR_HEADERS})''', bind=True
    )
    return await tab.wait_for_event(msg_id)


async def list_all_airports(tab: WSTab):
    return await make_ryanair_request(tab, f'{RYANAIR_API_URL}/views/locate/5/airports/en/active')


async def find_dest_airports(tab: WSTab, src_code: str):
    return await make_ryanair_request(tab, f'{RYANAIR_API_URL}/views/locate/searchWidget/routes/en/airport/{src_code}')


async def query_available_dates(tab: WSTab, src_code: str, dst_code: str):
    return await make_ryanair_request(tab, f'{RYANAIR_API_URL}/farfnd/3/oneWayFares/{src_code}/{dst_code}/availabilities')


async def query_flights_details(
    tab: WSTab, src_code: str, dst_code: str,
    date: datetime, days_before=2, days_after=2,
    adult=1, teen=0, child=0, infant=0
):
    date_str = date.strftime('%Y-%m-%d')
    return await make_ryanair_request(tab,
        f'{RYANAIR_API_URL}/booking/v4/en-gb/availability'
        f'?ADT={adult}&TEEN={teen}&CHD={child}&INF={infant}'
        f'&Origin={src_code}&Destination={dst_code}&promoCode=&IncludeConnectingFlights=false'
        f'&DateOut={date_str}&DateIn='
        f'&FlexDaysBeforeOut={days_before}&FlexDaysOut={days_after}'
        f'&FlexDaysBeforeIn={days_before}&FlexDaysIn={days_after}'
        f'&RoundTrip=false&IncludePrimeFares=false&ToUs=AGREED'
    )


async def make_worker(conn):
    ws_url = await conn.open_new_tab()
    websocket = await websockets.connect(ws_url).__aenter__()
    tab = WSTab(websocket)
    await tab.send_command("Page.enable")
    await tab.send_command('Page.navigate', url='https://www.ryanair.com/pl/pl/trip/flights/select?adults=1&teens=0&children=0&infants=0&dateOut=2025-10-14&dateIn=&isConnectedFlight=false&discount=0&promoCode=&isReturn=false&originIata=WAW&destinationIata=AGP&tpAdults=1&tpTeens=0&tpChildren=0&tpInfants=0&tpStartDate=2025-10-14&tpEndDate=&tpDiscount=0&tpPromoCode=&tpOriginIata=WAW&tpDestinationIata=AGP')
    await sleep_rand(4, 6)
    return tab


async def process_job(tab: WSTab, job: Job) -> Optional[Tuple[Job, ...]]:
    print('Processing job', job)
    if isinstance(job, QueryDatesJob):
        dates_data = await query_available_dates(tab, job.src_code, job.dst_code)
        logging.info('Available dates:', dates_data)
        dates = tuple(
            d for d in dates_data['dates'] if job.start_date <= d <= job.end_date
        )
        return tuple(make_dates_jobs(job.src_code, job.dst_code, dates))
    elif isinstance(job, QueryFlightsJob):
        date_parsed = datetime.strptime(job.date, '%Y-%m-%d')
        details_data = await query_flights_details(
            tab, job.src_code, job.dst_code, date_parsed,
            days_before=job.days_before, days_after=job.days_after
        )
        logging.info('Flight details:', details_data)
        await save_result('ryanair', details_data | {
            'type': 'details',
            'src_code': job.src_code,
            'dst_code': job.dst_code,
            'date': job.date
        })
    else:
        raise ValueError(f'Unknown job type: {job}')


async def download_ryanair(browser_conn, airports, start_date, end_date):
    try:
        jobs = load_jobs('ryanair_jobs.json', Job)
    except FileNotFoundError:
        jobs = tuple(init_jobs(airports, start_date, end_date))
    await run_jobs(
        jobs, process_job, partial(make_worker, browser_conn),
        max_workers=1, max_worker_jobs=1,
        save_jobs_fname='ryanair_jobs.json'
    )
